cmake: linker: separate toolchain_ld_options() calls for grouped flags

Fymyte opened the pull request:
Fixes #91045

Adding NO_SPLIT in the TOOLCHAIN_LD_FLAGS list caused every flags in the list to be checked together by the linker. This caused issues where incompatible flags was passed together, and the result was that none was propagated to the linker.

By using separate variables, we can achieve the initial goal of NO_SPLIT to only group a sub-set of flags.

Fymyte commented:
Hi @tejlmand, I was wondering if someone could review this MR before the 4.2 release. The underlying issue prevents us from linking because the default `-march/-mabi` flags from the toolchain is not what we want, but this is what's get used by Zephyr rn.

Fymyte commented:
Hi @tejlmand,
Hopefully I don't bother you during vacantion anymore.
Can you have a look at this MR ?
Thanks in advance

tejlmand reviewed (changes requested):
Thanks for this.

A couple of small observations, but principle looks sane :+1:

Fymyte committed:
cmake: linker: separate toolchain_ld_options() calls for grouped flags

Fixes #91045

Adding NO_SPLIT in the TOOLCHAIN_LD_FLAGS list caused every flags in the
list to be checked together by the linker. This caused issues where
incompatible flags was passed together, and the result was that none was
propagated to the linker.

By using separate variables, we can achieve the initial goal of NO_SPLIT to
only group a sub-set of flags.

Signed-off-by: Pierrick Guillaume <pguillaume@fymyte.com>

Fymyte commented:
Hi @tejlmand, do I need to do anything else ?

pdgendt commented:
Needs a rebase on `main`
